src/bookmarks/background/addition.js

```diff
@@ -57,18 +59,52 @@ export async function createNewPageForBookmark(id, bookmarkInfo) {
  * @throws {Error} Error thrown if page for supplied URL not indexed.
  * @returns {Promise<void>}
  */
-export async function createBookmarkByUrl(url) {
+export async function createBookmarkByUrl(url, tabId = null) {
     const pageId = generatePageDocId({ url })
-    const pageDoc = await db.get(pageId)
+    let pageDoc
+
+    // This is used for page is exist when we are bookmarking the page and default value is 1 that represents page exist in db.
+    let pageExist = 1
+
+    const timestamp = Date.now()
+
+    try {
+        pageDoc = await db.get(pageId)
+    } catch (err) {
+        console.log(err)
+        if (err.status === 404) {
+            pageExist = 0
+            const storePageResult = await storePage({ tabId, url })
+            pageDoc = (await storePageResult.finalPagePromise).page
+        }
+    }
 
     const bookmarkDoc = transformToBookmarkDoc({ _id: pageId })({
-        dateAdded: Date.now(),
+        dateAdded: timestamp,
         title: pageDoc.content.title,
         url: pageDoc.url,
     })
 
-    return await Promise.all([
-        index.addBookmarkConcurrent(pageId),
-        db.put(bookmarkDoc),
-    ])
+    if (pageExist) {
+        return await Promise.all([
+            index.addBookmarkConcurrent(pageId),
+            db.put(bookmarkDoc),
+        ])
+    } else {
+        // Create a visit pointing to this page (analysing/storing it may still be in progress)
```

Comment:
> In case page object doesnt exist > add page, visit and bookmark doc

Oliver says the same thing in the above comment.

Reply:
It should be handled in the follow-up log-page-visit call rather than here though. This logic is only concerned with the idea of bookmarking.

Remember the condition we added in the log-page-visit module was to decide between indexing page + visit, or only visit (if the page was already made for recent bookmark) - that logic still exists and should take care of that requirement. Remove this and see if that log-page-visit logic still covers this case